### ISSUE TITLE
How to hide or remove AWTGLCanvas in mac os? 

### ISSUE THREAD

**Issue by hubert-hy** [LWJGLX/lwjgl3-awt#38]
Hi, 

we implement AWTGLCanvas inside JPanel, and stack two JPanel with CardLayout. The issue is we need to hide the canvas when switching between the two panels. But so far I cannot do it. I tried `canvas.disposeCanvas()`, but it didn't work. Can you help me on that? Basically I need the `AWTGLCanvas` to go with its container, JPanel, to toggle the visibility. 

Thanks

Environment
LWJGL version: 3.0
LWJGL build #: X
Java version: 12
Platform: macOS

```java
import static org.lwjgl.opengl.GL.*;
import static org.lwjgl.opengl.GL11.*;
import java.awt.BorderLayout;
import java.awt.CardLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JLabel;
import javax.swing.JPanel;
public class CanvasTest {
	public static class LwjglPanel extends JPanel {
		public LwjglPanel(BorderLayout borderLayout) {
			super(borderLayout);
		}
		public void paint(Graphics g) {
			if (m_canvas != null)
				m_canvas.render();
		}
		@Override
		public void setVisible(boolean v) {
			super.setVisible(v);
			if (!v && m_canvas != null) {
				this.remove(m_canvas);
				this.m_canvas.disposeCanvas();
				this.m_canvas = null;
				this.repaint();
			}
		}
		public void init(AWTGLCanvas canvas) {
			this.m_canvas = canvas;
			this.add(m_canvas);
		}
		AWTGLCanvas m_canvas;
	}
	public static class CustomAWTGLCanvas extends AWTGLCanvas {
		public static GLData data;
		public static void create() {
			data = new GLData();
		}
		public CustomAWTGLCanvas() {
			super(data);
		}
		public void paintGL() {
			glClear(GL_COLOR_BUFFER_BIT);
			swapBuffers();
		}
		@Override
		public void initGL() {
			createCapabilities();
			glClearColor(0.3f, 0.4f, 0.5f, 1);
		}
	}
	public static void main(String[] args) {
		CustomAWTGLCanvas.create();
		CustomAWTGLCanvas canvas = new CustomAWTGLCanvas();
		JPanel panel = new LwjglPanel(new BorderLayout());
		panel.setMinimumSize(new Dimension(100, 100));
		((LwjglPanel) panel).init(canvas);
		JDialog frame = new JDialog();
		frame.setDefaultCloseOperation(JDialog.DISPOSE_ON_CLOSE);
		frame.setPreferredSize(new Dimension(640, 480));
		JPanel panels = new JPanel(new CardLayout());
		JButton switcher = new JButton("Switch Active Panel!");
		JPanel another = new JPanel();
		switcher.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				if (panel.isVisible()) {
					panel.setVisible(false);
					another.setVisible(true);
				} else {
					panel.setVisible(true);
					another.setVisible(false);
				}
			}
		});
		panels.add(panel);
		another.setPreferredSize(new Dimension(640, 480));
		another.setBackground(Color.GREEN);
		another.add(new JLabel("Yeah, this is another panel."));
		another.setVisible(false);
		panels.add(another);
		frame.add(switcher, BorderLayout.PAGE_START);
		frame.add(panels);
		frame.pack();
		frame.setVisible(true);
	}
}
```

**Comment by httpdigest** [comment]
I've edited your post (mostly: proper source code formatting and more concise example).

**Comment by httpdigest** [comment]
@tisoft can you reproduce, please?

**Comment by tisoft** [comment]
On macOS, the GLPanel is not removed.
On Linux, the GLPanel is removed, but it isn't displayed again, if you click the button again.

I can't test Windows. How does the example behave on Windows?

**Comment by httpdigest** [comment]
On Windows it does what the code says it does: Remove the panel once the button is clicked but not show it again.

**Comment by hubert-hy** [comment]
Thanks, I am right now only tested it on Mac, and do you think it could be a way to fix it? Before I use lwjgl 2's awtcanvas, but it broken after I upgrade to jdk 12. This new awtcanvas properly create opengl context with jdk 12, but seems it cannot hide with its container, here is the jpanel.

**Comment by httpdigest** [comment]
It has nothing to do with the JDK version... you probably also upgraded to LWJGL 3 (from LWJGL 2) in the process as well.

**Comment by hubert-hy** [comment]
Yes, I upgrade to lwjgl 3 and with lwjgl-awt, replace old lwjgl 2. Do you think this hiding issue can be fixed, I just tested in my test code, try to use disposeCanvas() to hide it, but no luck.

**Comment by tisoft** [comment]
With the fix in PR #39 setVisible works. No need for add/remove or dispose:

```
package org.lwjgl.opengl.awt;

import static org.lwjgl.opengl.GL.*;
import static org.lwjgl.opengl.GL11.*;
import java.awt.BorderLayout;
import java.awt.CardLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JDialog;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextArea;

public class CanvasTest {
	public static class LwjglPanel extends JPanel {
		public LwjglPanel(BorderLayout borderLayout) {
			super(borderLayout);
		}
		public void paint(Graphics g) {
			if (m_canvas != null)
				m_canvas.render();
		}
		public void init(AWTGLCanvas canvas) {
			this.m_canvas = canvas;
			this.add(m_canvas);
		}
		AWTGLCanvas m_canvas;
	}
	public static class CustomAWTGLCanvas extends AWTGLCanvas {
		public static GLData data;
		public static void create() {
			data = new GLData();
		}
		public CustomAWTGLCanvas() {
			super(data);
		}
		public void paintGL() {
			glClear(GL_COLOR_BUFFER_BIT);
			swapBuffers();
		}
		@Override
		public void initGL() {
			createCapabilities();
			glClearColor(0.3f, 0.4f, 0.5f, 1);
		}
	}
	public static void main(String[] args) {
		CustomAWTGLCanvas.create();
		CustomAWTGLCanvas canvas = new CustomAWTGLCanvas();
		JPanel panel = new LwjglPanel(new BorderLayout());
		panel.setMinimumSize(new Dimension(100, 100));
		((LwjglPanel) panel).init(canvas);
		JDialog frame = new JDialog();
		frame.setDefaultCloseOperation(JDialog.DISPOSE_ON_CLOSE);
		frame.setPreferredSize(new Dimension(640, 480));
		JPanel panels = new JPanel(new CardLayout());
		JButton switcher = new JButton("Switch Active Panel!");
		JPanel another = new JPanel();
		switcher.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				if (panel.isVisible()) {
					panel.setVisible(false);
					another.setVisible(true);
				} else {
					panel.setVisible(true);
					another.setVisible(false);
				}
			}
		});
		panels.add(panel);
		another.setPreferredSize(new Dimension(640, 480));
		another.setBackground(Color.GREEN);
		another.add(new JTextArea("Yeah, this is another panel."));
		another.setVisible(false);
		panels.add(another);
		frame.add(switcher, BorderLayout.PAGE_START);
		frame.add(panels);
		frame.pack();
		frame.setVisible(true);
	}
}
```

In fact removing does work, but re-adding does not. This is on all 3 platforms. But that is something for a different PR.

**Comment by hubert-hy** [comment]
Thanks, the fix is working.
